### ISSUE TITLE
Supporting Ammortized Inference for Normalizing Flows

### ISSUE THREAD

**Issue by gamerDecathlete** [pyro-ppl/pyro#1911]
### Issue Description (Feature request)
 I was going through the implementation for the PlanarFlows Module because it was bothering me how exactly they worked. I think the way they're set-up right now is....a bit incorrect? If you read the paper on it, the authors mention the parameters for the flow are produced by some function (it's like one sentence in the paper) and just to confirm it, I went to the code base for the sylvester flows paper and that's also what they do for their implementation of planar flows. 

### Code Snippet
This is just to highlight what I'm pointing out, it's the forward function for the planar flows in another repo. 

```
    def forward(self, zk, u, w, b):
        """
        Forward pass. Assumes amortized u, w and b. Conditions on diagonals of u and w for invertibility
        will be be satisfied inside this function. Computes the following transformation:
        z' = z + u h( w^T z + b)
        or actually
        z'^T = z^T + h(z^T w + b)u^T
        Assumes the following input shapes:
        shape u = (batch_size, z_size, 1)
        shape w = (batch_size, 1, z_size)
        shape b = (batch_size, 1, 1)
        shape z = (batch_size, z_size).
        """
```
This code snippets from [here](https://github.com/riannevdberg/sylvester-flows/blob/master/models/flows.py) line 33

**Comment by martinjankowiak** [comment]
yes, it's true that currently all our flows are not conditional, i.e. they cannot be conditioned on additional inputs. as such they do not directly support amortization. although the user is free to make the base distribution whatever he or she likes (in particular make them conditioned on data). as such these flows can still be used in the amortized context (see the DMM example). we are planning to make these flows more flexible by making them conditional. an initial version of this is probably forthcoming.

please reopen this issue if that doesn't answer your question.

**Comment by gamerDecathlete** [comment]
Thanks for the quick reply Martin. I'm a bit shaky with the terminology, but I think you've answered my question. If I understand you, what you're saying is currently with Pyro as is, I can't generate the parameters of the flow from a function like in the original paper. So, as these are all non conditional flows, then the parameters for the flows for any data I send through will be the same correct? I've read the DMM tutorial, and definitely I can still get the ammortization for the base distribution.  This is probably getting more into something for the discussion board, but supposed I really want to condition my flow parameters on my data because say...I'm looking to use pyro for my on going research and I already implemented a bunch of stuff with Pyro. What would you suggest for going about this? In my head, really everything is almost exactly the same, but in my guide instead of calling the current pyro Flow modules, I'm creating the flows each time. 
So in a rather pseudo code fashion:
```
fn guide (...):

  z_mean, z_sig = self.ammortized_basedistr(data)
   nf = []
   for i in self.num_flows:
           u = self.ammortized_u_i(data)
           w = self.ammortized_w_i(data)
           nf.append(ammortizedPlanarFlow(u, w) )
   ....
       with pyro.plate('data'):

                    dist = TransformedDistribution(Normal(z_mean, z_sig), nf) #WOULD THIS WORK?*

                z = pyro.sample('z', dist) 
```

Sorry if this is a bit confusing...but supposing I do something...like this, would the gradients for the ammortized_u_i and ammortize_w_i still work?

**Comment by martinjankowiak** [comment]
some version of that should work as long as you get all the shapes right.

until we merge conditional flows, the easiest thing to do would probably be to subclass the flow you
want (one that takes na `AutoregressiveNN` arg) and feed it a custom `ConditionalAutoregressiveNN` that you can condition on the data you want. this should be relatively straightforward but care needs to be taken to get all the tensor dimensions right.

@stefanwebb any comment?

**Comment by stefanwebb** [comment]
@gamerDecathlete thanks for your interest in the NF library! Yes, at the moment only joint distributions p(z) are supported and not yet conditional distributions p(z | x). Although this is what I'll be working on next so expect to have these features in the next few weeks.

To Martin's comment about your pseudocode, I'd also add that you have to make sure that the parameters of the NNs get added to the parameter store so that Pyro optimizes them. This would involve e.g. making `amortizedPlanarFlow` a `nn.Module` and saving `u`, `w` as `nn.Parameter` properties. Then call `pyro.module` on the flow object inside your model/guide.

Modifying `AutoregressiveNN` to make it conditional is a bit involved (this would be if you wanted to amortize e.g. IAF, NAF, etc. not planar flow). You would need to modify the MADE construction so that it takes some observed dims: https://github.com/pyro-ppl/pyro/blob/dev/pyro/nn/auto_reg_nn.py#L178. Then prepend the observed variable to the input in the forward function: https://github.com/pyro-ppl/pyro/blob/dev/pyro/nn/auto_reg_nn.py#L206.

**Comment by gamerDecathlete** [comment]
@stefanwebb thanks for the update! I probably won't get a chance to look into till next week but with all the input from both of you guys I think this seems fairly reasonable to implement with your guys advice. Thanks for also mentioning an approximate timeline on the Pyro dev for this feature. So if you finish in the next couple weeks I guess I should use...the nightly version of Pyro to use it right?  Either way, thanks for the feedback!
